contracts/strategy/yearn/YearnStrategy.sol

```diff
@@ -0,0 +1,166 @@
+        yVault = IYearnVault(_yVault);
+        underlying = IERC20(_underlying);
+    }
+
+    //
+    // Modifiers
+    //
+
+    modifier onlyManager() {
+        if (!hasRole(MANAGER_ROLE, msg.sender))
+            revert StrategyCallerNotManager();
+        _;
+    }
+
+    /// @inheritdoc IStrategy
+    function isSync() external pure override(IStrategy) returns (bool) {
+        return true;
+    }
+
+    /// @inheritdoc IStrategy
+    function hasAssets()
+        external
+        view
+        virtual
+        override(IStrategy)
+        returns (bool)
+    {
+        return _getShares() > 0;
+    }
+
+    /// @inheritdoc IStrategy
+    function investedAssets()
+        external
+        view
+        virtual
+        override(IStrategy)
+        returns (uint256)
+    {
+        return _sharesToUnderlying(_getShares());
+    }
+
+    /// @inheritdoc IStrategy
+    function invest() external virtual override(IStrategy) onlyManager {
+        uint256 balance = _getUnderlyingBalance();
+        if (balance == 0) revert StrategyNoUnderlying();
+
+        underlying.safeIncreaseAllowance(address(yVault), balance);
+
+        yVault.deposit(balance, address(this));
+    }
+
+    /// @inheritdoc IStrategy
+    function withdrawToVault(uint256 amount)
+        external
+        virtual
+        override(IStrategy)
+    {
+        if (amount == 0) revert StrategyAmountZero();
+        uint256 uninvestedUnderlying = _getUnderlyingBalance();
+
+        if (amount > uninvestedUnderlying) {
+            uint256 _sharesToWithdraw = _underlyingToShares(
+                amount - uninvestedUnderlying
+            );
+
+            if (_sharesToWithdraw > _getShares())
+                revert StrategyNotEnoughShares();
+
+            // burn shares and withdraw required underlying to strategy
+            yVault.withdraw(_sharesToWithdraw, address(this), 1);
+        }
+
+        // transfer underlying to vault
+        underlying.safeTransfer(vault, amount);
+    }
+
+    /**
+     * @dev Withdraws all the underlying to sandclock vault.
+     */
+    function withdrawAllToVault() external virtual onlyManager {
+        uint256 sharesBalance = _getShares();
+        if (sharesBalance == 0) revert StrategyNotEnoughShares();
+        // burn shares and withdraw required underlying to strategy
+        yVault.withdraw(sharesBalance, vault, 1);
+    }
+
+    /**
+     * @return LUSD balance of strategy
+     */
+    function _getUnderlyingBalance() internal view returns (uint256) {
+        return underlying.balanceOf(address(this));
+    }
+
+    /**
+     * @return yLusd balance of strategy
+     */
+    function _getShares() internal view returns (uint256) {
+        return yVault.balanceOf(address(this));
+    }
+
+    /**
+     * @return convert shares to lusd
+     */
+    function _sharesToUnderlying(uint256 _shares)
+        internal
+        view
+        returns (uint256)
+    {
+        return (_shares * yVault.pricePerShare()) / 1e18;
```

Comment:
not quite sure about this one, but should we first do division by `1e18` to avoid overflow situations?

Reply:
you usually do the multiplication first in order to avoid losing precision

Reply:
just an idea, how about we divide both `_shares` and `yVault.pricePerShare()` by `1e9` before they are multiplied? Does that make sense?